diff --git a/config.def.h b/config.def.h
index 4b07e6b..2048edc 100644
--- a/config.def.h
+++ b/config.def.h
@@ -18,6 +18,17 @@ static const char *colors[][3]      = {
 	[SchemeSel]  = { col_gray4, col_cyan,  col_cyan  },
 };
 
+static const TileMoveCallback tilemovefallback = NULL;
+
+static const int tilemovecfg
+    = TileMoveVertOverflow
+    | TileMoveHorzOverflow
+    | TileMoveMasterXDeny
+    | TileMoveMasterBottom
+    | TileMoveStackXDeny
+    | TileMoveStackTop
+    ;
+
 /* tagging */
 static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
 
@@ -70,6 +81,10 @@ static const Key keys[] = {
 	{ MODKEY,                       XK_k,      focusstack,     {.i = -1 } },
 	{ MODKEY,                       XK_i,      incnmaster,     {.i = +1 } },
 	{ MODKEY,                       XK_d,      incnmaster,     {.i = -1 } },
+    { MODKEY|ShiftMask,             XK_h,      tilemove,       {.i = 'W' } },
+    { MODKEY|ShiftMask,             XK_j,      tilemove,       {.i = 'S' } },
+    { MODKEY|ShiftMask,             XK_k,      tilemove,       {.i = 'N' } },
+    { MODKEY|ShiftMask,             XK_l,      tilemove,       {.i = 'E' } },
 	{ MODKEY,                       XK_h,      setmfact,       {.f = -0.05} },
 	{ MODKEY,                       XK_l,      setmfact,       {.f = +0.05} },
     { MODKEY,                       XK_x,      exstack,        {0} },
diff --git a/dwm.c b/dwm.c
index a51302b..e0ecd4b 100644
--- a/dwm.c
+++ b/dwm.c
@@ -67,6 +67,16 @@ enum { ClkTagBar, ClkLtSymbol, ClkStatusText, ClkWinTitle,
        ClkClientWin, ClkRootWin, ClkLast }; /* clicks */
 enum { XrayIn, XrayAc };
 enum { XrayGrpNone, XrayGrpTileMaster, XrayGrpTileStack };
+enum {
+    TileMoveVertOverflow = 1 << 0,
+    TileMoveHorzOverflow = 1 << 1,
+    TileMoveMasterTop    = 1 << 2,
+    TileMoveMasterBottom = 1 << 3,
+    TileMoveStackTop     = 1 << 4,
+    TileMoveStackBottom  = 1 << 5,
+    TileMoveMasterXDeny  = 1 << 6,
+    TileMoveStackXDeny   = 1 << 7,
+};
 
 typedef union {
 	int i;
@@ -159,6 +169,8 @@ typedef struct {
     int     st   ; /* is selected client in stack                            */
 } TileInfo;
 
+typedef void (*TileMoveCallback)(const Arg *);
+
 /* function declarations */
 static void applyrules(Client *c);
 static int applysizehints(Client *c, int *x, int *y, int *w, int *h, int interact);
@@ -233,6 +245,7 @@ static void tag(const Arg *arg);
 static void tagmon(const Arg *arg);
 static void tile(Monitor *m);
 static TileInfo tileinfo(void);
+static void tilemove(const Arg *arg);
 static void togglebar(const Arg *arg);
 static void togglefloating(const Arg *arg);
 static void toggletag(const Arg *arg);
@@ -1945,6 +1958,71 @@ tileinfo(void)
     return ti;
 }
 
+void
+tilemove(const Arg *arg)
+{
+    TileInfo ti;
+    Client *c, **bsts[2] = { ti.bs, ti.ts };
+    int mm, mx, ms, mt, north, south, west, east, of, dir;
+
+    if (selmon->lt[selmon->sellt]->arrange != &tile) {
+        if (tilemovefallback)
+            tilemovefallback(arg);
+        return;
+    }
+
+    if (!arg || !selmon->sel || selmon->sel->isfloating ||
+            (selmon->sel->isfullscreen && lockfullscreen))
+        return;
+
+    mm = tilemovecfg & (TileMoveMasterTop | TileMoveMasterBottom);
+    ms = tilemovecfg & (TileMoveStackTop  | TileMoveStackBottom );
+    if ( (mm != TileMoveMasterTop && mm != TileMoveMasterBottom)
+      || (ms != TileMoveStackTop  && ms != TileMoveStackBottom ) )
+        return;
+
+    ti = tileinfo();
+    if (ti.nc == 0)
+        return;
+
+    north = arg->i == 'N' || arg->i == 'n';
+    south = arg->i == 'S' || arg->i == 's';
+    mx    = tilemovecfg & (ti.st ? TileMoveStackXDeny : TileMoveMasterXDeny);
+    if (north || south) {
+        if (mx && ti.xs[ti.st])
+            return;
+        c = (bsts[north][ti.st] == selmon->sel)
+          ? (tilemovecfg & TileMoveVertOverflow)
+          ? bsts[south][ti.st] : NULL : ti.cs[south];
+        if (c) {
+            cswap(c, selmon->sel);
+            arrange(selmon);
+        }
+        return;
+    }
+
+    west = arg->i == 'W' || arg->i == 'w';
+    east = arg->i == 'E' || arg->i == 'e';
+    of   = tilemovecfg & TileMoveHorzOverflow;
+    dir  = ( ti.st && west) ? 0
+         : (!ti.st && west) ? (of ? 1 : -1)
+         : ( ti.st && east) ? (of ? 0 : -1)
+         : (!ti.st && east) ? 1
+         : -1
+         ;
+
+    if (!(west || east) || dir == -1)
+        return;
+
+    selmon->nmaster = dir
+        ? MIN(selmon->nmaster, ti.nc) - 1
+        : selmon->nmaster + 1;
+    mt = (tilemovecfg & (dir ? TileMoveStackTop : TileMoveMasterTop)) != 0;
+    c  = bsts[mt][dir];
+    cmove(selmon->sel, c ? c : bsts[!dir][!dir], c ? !mt : dir);
+    arrange(selmon);
+}
+
 void
 togglebar(const Arg *arg)
 {
