diff --git a/config.def.h b/config.def.h
index 81c3fc0..24f9fee 100644
--- a/config.def.h
+++ b/config.def.h
@@ -39,10 +39,16 @@ static const int lockfullscreen = 1; /* 1 will force focus on the fullscreen win
 static const int refreshrate = 120;  /* refresh rate (per second) for client move/resize */
 
 static const Layout layouts[] = {
-	/* symbol     arrange function */
-	{ "[]=",      tile },    /* first entry is default */
-	{ "><>",      NULL },    /* no layout function means floating behavior */
-	{ "[M]",      monocle },
+	/*               symbol     arrange function */
+	[LtTile]    = { "[]=",      tile    }, /* first entry is default */
+	[LtFloat]   = { "><>",      NULL    }, /* no layout function means floating behavior */
+	[LtMonocle] = { "[M]",      monocle },
+};
+
+static const Layout *layoutring[] = {
+    &layouts[LtTile],
+    &layouts[LtMonocle],
+    NULL
 };
 
 /* key definitions */
@@ -80,6 +86,7 @@ static const Key keys[] = {
 	{ MODKEY,                       XK_m,      setlayout,      {.v = &layouts[2]} },
 	{ MODKEY,                       XK_space,  setlayout,      {0} },
 	{ MODKEY|ShiftMask,             XK_space,  togglefloating, {0} },
+    { MODKEY,                       XK_c,      cyclebetween,   {.v = layoutring } },
 	{ MODKEY,                       XK_0,      view,           {.ui = ~0 } },
 	{ MODKEY|ShiftMask,             XK_0,      tag,            {.ui = ~0 } },
 	{ MODKEY,                       XK_comma,  focusmon,       {.i = -1 } },
diff --git a/dwm.c b/dwm.c
index 53b393e..6be911f 100644
--- a/dwm.c
+++ b/dwm.c
@@ -58,6 +58,7 @@
 
 /* enums */
 enum { CurNormal, CurResize, CurMove, CurLast }; /* cursor */
+enum { LtTile, LtFloat, LtMonocle }; /* layouts */
 enum { SchemeNorm, SchemeSel }; /* color schemes */
 enum { NetSupported, NetWMName, NetWMState, NetWMCheck,
        NetWMFullscreen, NetActiveWindow, NetWMWindowType,
@@ -156,6 +157,7 @@ static void configure(Client *c);
 static void configurenotify(XEvent *e);
 static void configurerequest(XEvent *e);
 static Monitor *createmon(void);
+static void cyclebetween(const Arg *);
 static void destroynotify(XEvent *e);
 static void detach(Client *c);
 static void detachstack(Client *c);
@@ -646,6 +648,25 @@ createmon(void)
 	return m;
 }
 
+void
+cyclebetween(const Arg *arg)
+{
+    Layout **i;
+    Arg lt;
+
+    if (!(arg && arg->v))
+        return;
+
+    for (i = (Layout **)arg->v; *i; ++i)
+        if (selmon->lt[selmon->sellt] == *i)
+            break;
+
+    lt.v = *i && i[1]
+         ? i[1]
+         : *(Layout **)(arg->v);
+    setlayout(&lt);
+}
+
 void
 destroynotify(XEvent *e)
 {
