diff --git a/config.def.h b/config.def.h
index 81c3fc0..4b07e6b 100644
--- a/config.def.h
+++ b/config.def.h
@@ -72,6 +72,7 @@ static const Key keys[] = {
 	{ MODKEY,                       XK_d,      incnmaster,     {.i = -1 } },
 	{ MODKEY,                       XK_h,      setmfact,       {.f = -0.05} },
 	{ MODKEY,                       XK_l,      setmfact,       {.f = +0.05} },
+    { MODKEY,                       XK_x,      exstack,        {0} },
 	{ MODKEY,                       XK_Return, zoom,           {0} },
 	{ MODKEY,                       XK_Tab,    view,           {0} },
 	{ MODKEY|ShiftMask,             XK_c,      killclient,     {0} },
diff --git a/dwm.c b/dwm.c
index 3fc3fb3..4e8c806 100644
--- a/dwm.c
+++ b/dwm.c
@@ -66,7 +66,7 @@ enum { WMProtocols, WMDelete, WMState, WMTakeFocus, WMLast }; /* default atoms *
 enum { ClkTagBar, ClkLtSymbol, ClkStatusText, ClkWinTitle,
        ClkClientWin, ClkRootWin, ClkLast }; /* clicks */
 enum { XrayIn, XrayAc };
-enum { XrayGrpNone };
+enum { XrayGrpNone, XrayGrpTileMaster, XrayGrpTileStack };
 
 typedef union {
 	int i;
@@ -101,6 +101,11 @@ struct Client {
 	Window win;
 };
 
+typedef struct {
+    int master;
+    int stack;
+} ExStack;
+
 typedef struct {
 	unsigned int mod;
 	KeySym keysym;
@@ -132,6 +137,7 @@ struct Monitor {
 	Monitor *next;
 	Window barwin;
 	const Layout *lt[2];
+    ExStack stacks;
 };
 
 typedef struct {
@@ -167,6 +173,7 @@ static void drawbar(Monitor *m);
 static void drawbars(void);
 static void enternotify(XEvent *e);
 static void expose(XEvent *e);
+static void exstack(const Arg *arg);
 static void focus(Client *c);
 static void focusin(XEvent *e);
 static void focusmon(const Arg *arg);
@@ -280,6 +287,8 @@ static Monitor *mons, *selmon;
 static Window root, wmcheckwin;
 static void (*xrayhandlers[])(Client *) = {
     [XrayGrpNone] = NULL,
+    [XrayGrpTileMaster] = xraydefhandler,
+    [XrayGrpTileStack]  = xraydefhandler,
 };
 
 /* configuration, allows nested code to access above variables */
@@ -664,6 +673,8 @@ createmon(void)
 	m->topbar = topbar;
 	m->lt[0] = &layouts[0];
 	m->lt[1] = &layouts[1 % LENGTH(layouts)];
+    m->stacks.master = 0;
+    m->stacks.stack  = 0;
 	strncpy(m->ltsymbol, layouts[0].symbol, sizeof m->ltsymbol);
 	return m;
 }
@@ -807,6 +818,26 @@ expose(XEvent *e)
 		drawbar(m);
 }
 
+void
+exstack(const Arg *arg)
+{
+    Client *c;
+    int n;
+
+    if (selmon->lt[selmon->sellt]->arrange != &tile
+            || !selmon->sel || selmon->sel->isfloating)
+        return;
+
+    for (n = 0, c = nexttiled(selmon->clients); c && selmon->sel != c; c = nexttiled(c->next), ++n);
+
+    if (n < selmon->nmaster)
+        selmon->stacks.master = !selmon->stacks.master;
+    else
+        selmon->stacks.stack = !selmon->stacks.stack;
+
+    arrange(selmon);
+}
+
 void
 focus(Client *c)
 {
@@ -1714,29 +1745,58 @@ tagmon(const Arg *arg)
 void
 tile(Monitor *m)
 {
-	unsigned int i, n, h, mw, my, ty;
-	Client *c;
+    unsigned int i, n, h, mw, my, ty, si, ma, sa;
+    Client *c, *xr[2];
 
-	for (n = 0, c = nexttiled(m->clients); c; xraycfg(c, XrayGrpNone, -1), c = nexttiled(c->next), n++);
-	if (n == 0)
-		return;
+    for (n = si = 0, xr[0] = xr[1] = NULL, c = nexttiled(m->clients); c; c = nexttiled(c->next), ++n) {
+        if (m->sel == c)
+            si = n;
 
-	if (n > m->nmaster)
-		mw = m->nmaster ? m->ww * m->mfact : 0;
-	else
-		mw = m->ww;
-	for (i = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++)
-		if (i < m->nmaster) {
-			h = (m->wh - my) / (MIN(n, m->nmaster) - i);
-			resize(c, m->wx, m->wy + my, mw - (2*c->bw), h - (2*c->bw), 0);
-			if (my + HEIGHT(c) < m->wh)
-				my += HEIGHT(c);
-		} else {
-			h = (m->wh - ty) / (n - i);
-			resize(c, m->wx + mw, m->wy + ty, m->ww - mw - (2*c->bw), h - (2*c->bw), 0);
-			if (ty + HEIGHT(c) < m->wh)
-				ty += HEIGHT(c);
-		}
+        if (c->xraystat == XrayAc) switch (c->xraygrp) {
+            case XrayGrpTileMaster : xr[0] = c; break;
+            case XrayGrpTileStack  : xr[1] = c; break;
+        }
+    }
+
+    if (n == 0)
+        return;
+
+    if (m->sel && !m->sel->isfloating)
+        xr[si >= m->nmaster] = m->sel;
+
+    if (n > m->nmaster)
+        mw = m->nmaster ? m->ww * m->mfact : 0;
+    else
+        mw = m->ww;
+    for (i = ma = sa = my = ty = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), i++) {
+        if (i < m->nmaster) {
+            xraycfg( c
+                   , m->stacks.master ? XrayGrpTileMaster : XrayGrpNone
+                   , xr[0] == c ? (ma = 1, XrayAc) : XrayIn
+                   );
+            h = m->stacks.master ? m->wh : (m->wh - my) / (MIN(n, m->nmaster) - i);
+            resize(c, m->wx, m->wy + my * !m->stacks.master, mw - (2*c->bw), h - (2*c->bw), 0);
+            if (my + HEIGHT(c) < m->wh)
+                my += HEIGHT(c);
+        } else {
+            xraycfg( c
+                   , m->stacks.stack ? XrayGrpTileStack : XrayGrpNone
+                   , xr[1] == c ? (sa = 1, XrayAc) : XrayIn
+                   );
+            h = m->stacks.stack ? m->wh : (m->wh - ty) / (n - i);
+            resize(c, m->wx + mw, m->wy + ty * !m->stacks.stack, m->ww - mw - (2*c->bw), h - (2*c->bw), 0);
+            if (ty + HEIGHT(c) < m->wh)
+                ty += HEIGHT(c);
+        }
+
+        xrayapply(c);
+    }
+
+    if (!ma && m->stacks.master)
+        xrayfallback(XrayGrpTileMaster);
+
+    if (!sa && m->stacks.stack)
+        xrayfallback(XrayGrpTileStack);
 }
 
 void
