diff --git a/dwm.c b/dwm.c
index 53b393e..3fc3fb3 100644
--- a/dwm.c
+++ b/dwm.c
@@ -65,6 +65,8 @@ enum { NetSupported, NetWMName, NetWMState, NetWMCheck,
 enum { WMProtocols, WMDelete, WMState, WMTakeFocus, WMLast }; /* default atoms */
 enum { ClkTagBar, ClkLtSymbol, ClkStatusText, ClkWinTitle,
        ClkClientWin, ClkRootWin, ClkLast }; /* clicks */
+enum { XrayIn, XrayAc };
+enum { XrayGrpNone };
 
 typedef union {
 	int i;
@@ -92,6 +94,7 @@ struct Client {
 	int bw, oldbw;
 	unsigned int tags;
 	int isfixed, isfloating, isurgent, neverfocus, oldstate, isfullscreen;
+    int xraygrp, xraystat;
 	Client *next;
 	Client *snext;
 	Monitor *mon;
@@ -231,6 +234,15 @@ static Monitor *wintomon(Window w);
 static int xerror(Display *dpy, XErrorEvent *ee);
 static int xerrordummy(Display *dpy, XErrorEvent *ee);
 static int xerrorstart(Display *dpy, XErrorEvent *ee);
+static Client *xrayac(Monitor *mon, int grp);
+static Client *xrayacs(Monitor *mon, int grp, Client *skip);
+static void xrayapply(Client *c);
+static void xraycfg(Client *c, int grp, int stat);
+static void xraydefhandler(Client *c);
+static void xrayfallback(int grp);
+static void xrayfocus(Client *c);
+static void xraymaybefallback(int grp);
+static void xrayunfocus(Client *c);
 static void zoom(const Arg *arg);
 
 /* variables */
@@ -266,6 +278,9 @@ static Display *dpy;
 static Drw *drw;
 static Monitor *mons, *selmon;
 static Window root, wmcheckwin;
+static void (*xrayhandlers[])(Client *) = {
+    [XrayGrpNone] = NULL,
+};
 
 /* configuration, allows nested code to access above variables */
 #include "config.h"
@@ -395,9 +410,16 @@ arrange(Monitor *m)
 void
 arrangemon(Monitor *m)
 {
+    Client *c;
+
 	strncpy(m->ltsymbol, m->lt[m->sellt]->symbol, sizeof m->ltsymbol);
-	if (m->lt[m->sellt]->arrange)
-		m->lt[m->sellt]->arrange(m);
+    if (m->lt[m->sellt]->arrange)
+        m->lt[m->sellt]->arrange(m);
+    else
+        for (c = nexttiled(m->clients); c; c = nexttiled(c->next)) {
+            xraycfg(c, XrayGrpNone, -1);
+            xrayapply(c);
+        }
 }
 
 void
@@ -790,6 +812,7 @@ focus(Client *c)
 {
 	if (!c || !ISVISIBLE(c))
 		for (c = selmon->stack; c && !ISVISIBLE(c); c = c->snext);
+    xrayfocus(c);
 	if (selmon->sel && selmon->sel != c)
 		unfocus(selmon->sel, 0);
 	if (c) {
@@ -1037,6 +1060,7 @@ manage(Window w, XWindowAttributes *wa)
 
 	c = ecalloc(1, sizeof(Client));
 	c->win = w;
+    xraycfg(c, XrayGrpNone, -1);
 	/* geometry */
 	c->x = c->oldx = wa->x;
 	c->y = c->oldy = wa->y;
@@ -1121,8 +1145,10 @@ monocle(Monitor *m)
 			n++;
 	if (n > 0) /* override layout symbol */
 		snprintf(m->ltsymbol, sizeof m->ltsymbol, "[%d]", n);
-	for (c = nexttiled(m->clients); c; c = nexttiled(c->next))
+	for (c = nexttiled(m->clients); c; c = nexttiled(c->next)) {
+        xraycfg(c, XrayGrpNone, -1);
 		resize(c, m->wx, m->wy, m->ww - 2 * c->bw, m->wh - 2 * c->bw, 0);
+    }
 }
 
 void
@@ -1490,6 +1516,7 @@ setfullscreen(Client *c, int fullscreen)
 		c->oldbw = c->bw;
 		c->bw = 0;
 		c->isfloating = 1;
+        xraycfg(c, XrayGrpNone, -1);
 		resizeclient(c, c->mon->mx, c->mon->my, c->mon->mw, c->mon->mh);
 		XRaiseWindow(dpy, c->win);
 	} else if (!fullscreen && c->isfullscreen){
@@ -1690,7 +1717,7 @@ tile(Monitor *m)
 	unsigned int i, n, h, mw, my, ty;
 	Client *c;
 
-	for (n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), n++);
+	for (n = 0, c = nexttiled(m->clients); c; xraycfg(c, XrayGrpNone, -1), c = nexttiled(c->next), n++);
 	if (n == 0)
 		return;
 
@@ -1729,9 +1756,11 @@ togglefloating(const Arg *arg)
 	if (selmon->sel->isfullscreen) /* no support for fullscreen windows */
 		return;
 	selmon->sel->isfloating = !selmon->sel->isfloating || selmon->sel->isfixed;
-	if (selmon->sel->isfloating)
+	if (selmon->sel->isfloating) {
+        xraycfg(selmon->sel, XrayGrpNone, -1);
 		resize(selmon->sel, selmon->sel->x, selmon->sel->y,
 			selmon->sel->w, selmon->sel->h, 0);
+    }
 	arrange(selmon);
 }
 
@@ -1773,6 +1802,7 @@ unfocus(Client *c, int setfocus)
 		XSetInputFocus(dpy, root, RevertToPointerRoot, CurrentTime);
 		XDeleteProperty(dpy, root, netatom[NetActiveWindow]);
 	}
+    xrayunfocus(c);
 }
 
 void
@@ -2128,6 +2158,118 @@ xerrorstart(Display *dpy, XErrorEvent *ee)
 	return -1;
 }
 
+Client *
+xrayac(Monitor *mon, int grp)
+{
+    return xrayacs(mon, grp, NULL);
+}
+
+Client *
+xrayacs(Monitor *mon, int grp, Client *skip)
+{
+    Client *c;
+
+    if (!mon)
+        return NULL;
+
+    for (c = mon->clients; c; c = c->next)
+        if (c != skip && ISVISIBLE(c) && c->xraygrp == grp && c->xraystat == XrayAc)
+            break;
+
+    return c;
+}
+
+void
+xrayapply(Client *c)
+{
+    if (c && xrayhandlers[c->xraygrp])
+        xrayhandlers[c->xraygrp](c);
+}
+
+void
+xraycfg(Client *c, int grp, int stat)
+{
+    if (!c)
+        return;
+
+    if (grp == XrayGrpNone) {
+        c->xraygrp  = XrayGrpNone;
+        c->xraystat = 0;
+        return;
+    }
+
+    c->xraygrp  = grp;
+    c->xraystat = stat;
+}
+
+void
+xraydefhandler(Client *c)
+{
+    switch (c->xraystat) {
+        case XrayIn:
+            XMoveWindow(dpy, c->win, WIDTH(c) * -2, c->y);
+            break;
+        case XrayAc:
+            XMoveWindow(dpy, c->win, c->x, c->y);
+            break;
+    }
+}
+
+void
+xrayfallback(int grp)
+{
+    Client *c;
+
+    if (grp == XrayGrpNone)
+        return;
+
+    for (c = selmon->stack; c; c = c->snext)
+        if (ISVISIBLE(c) && c->xraygrp == grp) {
+            xraycfg(c, grp, XrayAc);
+            xrayapply(c);
+            return;
+        }
+}
+
+void
+xrayfocus(Client *c)
+{
+    Client *i;
+
+    if (!c || c->xraygrp == XrayGrpNone)
+        return;
+
+    xraycfg(c, c->xraygrp, XrayAc);
+    xrayapply(c);
+    for (i = c->mon->clients; i; i = i->next)
+        if (i != c && ISVISIBLE(i) && i->xraygrp == c->xraygrp && i->xraystat == XrayAc) {
+            xraycfg(i, i->xraygrp, XrayIn);
+            xrayapply(i);
+        }
+}
+
+void
+xraymaybefallback(int grp)
+{
+    if (grp == XrayGrpNone)
+        return;
+
+    if (!xrayac(selmon, grp))
+        xrayfallback(grp);
+}
+
+void
+xrayunfocus(Client *c)
+{
+    if (!c || c->xraygrp == XrayGrpNone)
+        return;
+
+    if (xrayacs(c->mon, c->xraygrp, c)) {
+        xraycfg(c, c->xraygrp, XrayIn);
+        xrayapply(c);
+    }
+}
+
 void
 zoom(const Arg *arg)
 {
